test/runtests.jl

```diff
@@ -68,6 +68,120 @@ end
                    8.848886704358202, 2.8697881579099085, 11.15107375630744, 11.8392578374927])
 end
 
+@testset "rankdeficient-glm" begin
+    Random.seed!(9999)
+    df = DataFrames.DataFrame()
+    num_rows = 100
+    df[:feature_1] = randn(num_rows)
+    df[:feature_2] = 2 .* df[:feature_1]
+    df[:feature_3] = randn(num_rows)
+    df[:label_1] = round.(Int, rand(num_rows))
+
+    Random.seed!(9999)
+    OLS_1 = GLM.lm(
+        StatsModels.@formula(
+            label_1 ~ 1 + feature_1 + feature_2 + feature_3
+            ),
+        df,
+        true,
+        )
+    Test.@test(0.4 < GLM.coeftable(OLS_1.model).cols[1][1] < 0.6)
+    Test.@test(isapprox(GLM.coeftable(OLS_1.model).cols[1][2], 0.0))
+    Test.@test(0.003 < GLM.coeftable(OLS_1.model).cols[1][3] < 0.004)
+    Test.@test(0.004 < GLM.coeftable(OLS_1.model).cols[1][4] < 0.005)
+    Test.@test(GLM.coeftable(OLS_1.model).cols[4][1].v < 1e-10)
+    Test.@test(isnan(GLM.coeftable(OLS_1.model).cols[4][2].v))
+    Test.@test(0.8 < GLM.coeftable(OLS_1.model).cols[4][3].v < 1.0)
+    Test.@test(0.9 < GLM.coeftable(OLS_1.model).cols[4][4].v < 1.0)
+
+    Random.seed!(9999)
+    Test.@test_throws(
+        LinearAlgebra.PosDefException,
+        GLM.glm(
+            StatsModels.@formula(
+                label_1 ~ 1 + feature_1 + feature_2 + feature_3
+                ),
+            df,
+            GLM.Binomial();
+            allowrankdeficient = false,
+            ),
+        )
+
+    Random.seed!(9999)
+    LR_1 = GLM.glm(
+        StatsModels.@formula(
+            label_1 ~ 1 + feature_1 + feature_2 + feature_3
+            ),
+        df, GLM.Binomial();
+        allowrankdeficient = true,
+        )
+    Test.@test(-0.1 < GLM.coeftable(LR_1.model).cols[1][1] < 0.0)
+    Test.@test(GLM.coeftable(LR_1.model).cols[1][2] < -4e12)
+    Test.@test(GLM.coeftable(LR_1.model).cols[1][3] > 2e12)
+    Test.@test(0.01 < GLM.coeftable(LR_1.model).cols[1][4] < 0.03)
+    Test.@test(0.6 < GLM.coeftable(LR_1.model).cols[4][1].v < 0.8)
+    Test.@test(isnan(GLM.coeftable(LR_1.model).cols[4][2].v))
+    Test.@test(GLM.coeftable(LR_1.model).cols[4][3].v < 1e-90)
+    Test.@test(0.9 < GLM.coeftable(LR_1.model).cols[4][4].v < 1.0)
+
+    Random.seed!(123)
+    df[:label_2] = Vector{Int}(undef, num_rows)
```

Comment:
Suggestion for clarity,
```
df.label_2 = ifelse.(df.feature_3 .> 0,
                     sample(0:1, weights([0.2, 0.8])),
                     sample(0:1, weights([0.8, 0.2])))
```
or
```
df.label_2 = rand.(Bernoulli.(ifelse.(df.feature_3 .> 0, 0.2, 0.8)))
```

Reply:
I'm going to read the test data from file, instead of generating it randomly. That way, the tests will be deterministic and should always pass.